substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs

```diff
@@ -1395,6 +1395,84 @@ where
 		}
 	}
 
+	/// Fetches the 'provides' tags associated to transactions part of blocks
+	/// of an enacted fork.
+	///
+	/// The 'provides' tags of transactions from enacted blocks are searched
+	/// in inactive views. These tags serve as cache based on the inactive views,
+	/// since prunning transactions is based on their 'provides' tags, which if
+	/// not found in the enacted fork tip active view, then they will be obtained
+	/// by revalidating the transaction, which takes time.
+	async fn provides_tags_from_inactive_views(
+		&self,
+		tree_route: &TreeRoute<Block>,
+	) -> HashMap<ExtrinsicHash<ChainApi>, Vec<Tag>> {
+		// For every enacted block, take its txs.
+		let blocks_xts = future::join_all(
+			tree_route
+				.enacted()
+				.iter()
+				.map(|hn| {
+				    let api = self.api.clone();
+					async move {
+						api
+							.block_body(hn.hash)
+							.await
+							.unwrap_or_else(|e| {
+								log::warn!(target: LOG_TARGET, "provides_tags_from_inactive_views: block_body error request: {}", e);
+								None
+							})
+							.unwrap_or_default()
+					}
+				}))
+			.await
+			.into_iter()
+			.collect::<Vec<_>>();
```

Comment:
We could flatten and collect all txs into single view. Is there a reason to have them grouped by block?

Reply:
No reason, just haven't thought this through fully. Txs part of different blocks should be disjunct too, so no need to worry about intersections I think. I will flatten.